### PR TITLE
[reland] Call jit decomposition in VariableType to increase forward AD coverage (#84151)

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #84675

This reverts commit acb4a09628284201281e262aaee58e3dc6be9c2b.

In addition, we also fix a memory leak in layer norm.

### CODE REVIEW

File: torch/_decomp/decompositions_for_jvp.py [hunk 1]
```diff
@@ -179,7 +179,7 @@ def native_layer_norm_backward(
         if len(outer_dim_indices) > 0:
             d_bias: Optional[Tensor] = torch.sum(grad_out, outer_dim_indices, False)
         else:
-            d_bias = grad_out
+            d_bias = grad_out.clone()
```
Comment (soulitzer) [inline torch/_decomp/decompositions_for_jvp.py:182]: This is the only thing that changed

Comment (soulitzer) [reply torch/_decomp/decompositions_for_jvp.py:182]: In combination with the changes in this PR, returning this input as-is as an output makes it so that we create a cycle in the graph.

Comment (zou3519) [reply torch/_decomp/decompositions_for_jvp.py:182]: Can you explain the memory leak a bit more?

It would be concerning if we used decompositions for operators that didn't have forward-mode AD formulas by default and if this is an easy way to introduce memory leaks

Comment (soulitzer) [reply torch/_decomp/decompositions_for_jvp.py:182]: Suppose we have the function F(a) -> a (which just returns the input as-is) and we pass a leaf variable that that function, after the function completes we will have the cycle:

`a-> FBackward -> a's accumulate grad -> a`

This is why we don't allow the input tensor to be returned as-is for a non-composite functions (the TensorImpl use count checks should've helped with that here)

I don't think it is necessary for decompositions to follow ALL the same rules that a non-composite kernel should follow though. For example, the rules regarding views don't apply because the ADInplaceOverView is actually done after the decomposition so returning a view only sometimes should be OK.

File: tools/autograd/gen_variable_type.py [hunk 2]
```diff
@@ -601,7 +601,7 @@
   if (${try_jit_decomposition_bool} && ${any_has_forward_grad}) {
     static c10::OperatorName full_name("aten::${op_name}", "${op_overload}");
     static c10::optional<c10::OperatorHandle> opt_op = c10::Dispatcher::singleton().findSchema(full_name);
-    return impl::run_jit_decomposition_with_args_for_jvp<${returns_and_args}>("${op_name}", *opt_op, ks, ${arg_names});
+    return impl::run_jit_decomposition_with_args_for_jvp<${return_types}>("${op_name}", *opt_op, ks, ${arg_names});
```
Comment (soulitzer) [inline tools/autograd/gen_variable_type.py:604]: Updated to only pass in return type here and let c++ infer the rest

File: tools/autograd/derivatives.yaml [hunk 3]
```diff
@@ -1956,7 +1956,20 @@
...
+  # In functorch we rely on OP_DECOMPOSE, but now we compute forward AD using an actual
+  # formula.
+  #
+  # We'd like to avoid keeping the entire jvp decomposition mechanism in functorch,
+  # just for this single decomposition, but also want to avoid any cases from regressing:
+  # e.g. test_vmapjvpall_nn_functional_logsigmoid_cuda_float32 (passes on cpu, fails on CUDA).
+  #
+  # We should either figure out what is going on with vmap or perhaps fwd AD could
+  # be more tolerant about 0-dim vs 1-dim tensors
+  output: log_sigmoid_backward(self_t.conj(), self_p, buffer).conj().view_as(self_p)
```
Comment (zou3519) [inline tools/autograd/derivatives.yaml:1972]: It sounds like otherwise we will regress `test_vmapjvpall_nn_functional_logsigmoid_cuda_float32`. Does this mean that `auto_element_wise` is incorrect (unlikely), or there's a bug in vmap or forward AD?

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -3323,7 +3327,7 @@ def sample_inputs_layer_norm(opinfo, device, dtype, requires_grad, **kwargs):
             kwargs=kwargs
         )
     # Without any optional args
-    yield SampleInput(make_arg((1, 2)), args=((2,),))
+    # yield SampleInput(make_arg((1, 2)), args=((2,),))
```
Comment (zou3519) [inline torch/testing/_internal/common_methods_invocations.py:3330]: What's going on here?

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -39,6 +39,10 @@
 import torch._refs.special
 import torch._refs.linalg
 
+# Make sure that decompositions used for test_forward_mode_AD and
+# test_fn_fwgrad_bwgrad are registered to the jit
+import torch._decomp.decompositions_for_jvp
+
```
Comment (zou3519) [inline torch/testing/_internal/common_methods_invocations.py:45]: Hold on: when do these decompositions get registered outside of testing? It's a bit weird to require users to call `import torch._decomp.decompositions_for_jvp` before they can use the decompositions

File: torch/_decomp/decompositions_for_jvp.py [hunk 6]
```diff
@@ -179,7 +179,7 @@ def native_layer_norm_backward(
         if len(outer_dim_indices) > 0:
             d_bias: Optional[Tensor] = torch.sum(grad_out, outer_dim_indices, False)
         else:
-            d_bias = grad_out
+            d_bias = grad_out.view_as(grad_out)
     elif bias is not None:
```
Comment (zou3519) [inline torch/_decomp/decompositions_for_jvp.py:183]: What is going on in line 26?
```
# TODO: what if jit decompositions exists, should we just use it?
#       or do we want to have an explicit white list like functorch had
#       using special JVP_DECOMP DynamicLayerFront kernel
```

File: torch/_decomp/decompositions_for_jvp.py [hunk 7]
```diff
@@ -179,7 +179,7 @@ def native_layer_norm_backward(
         if len(outer_dim_indices) > 0:
             d_bias: Optional[Tensor] = torch.sum(grad_out, outer_dim_indices, False)
         else:
-            d_bias = grad_out
+            d_bias = grad_out.view_as(grad_out)
```
Comment (zou3519) [inline torch/_decomp/decompositions_for_jvp.py:182]: I have two questions:
1. `native_layer_norm_backward`'s decomposition only gets registered to `decomposition_table_for_jvp`. Is there a reason why we didn't put it into torch._decomps.decomposition_table? @Chillee 
2. Claim: a decomposition for an operator should match the alias information in the schema. Because the alias information for native_layer_norm_backward explicitly says that `grad_out` is not aliased, then we should be returning a clone here, otherwise we are lying.

File: torch/_decomp/decompositions_for_jvp.py (hunk 1 above)
Comment (zou3519) [reply torch/_decomp/decompositions_for_jvp.py]: > Suppose we have the function F(a) -> a (which just returns the input as-is) and we pass a leaf variable that that function, after the function completes we will have the cycle:

Thanks for the explanation, that makes sense

**Review by zou3519 — COMMENTED** [review]
Added some questions/comments, but to summarize there's at least two things I think we should resolve before landing this:
1. the decomposition should probably match the schema (and not do a view), but I could be convinced otherwise
2. auto_element_wise for log_sigmoid_forward not working is very suspicious

(2) could be handled in a follow-up if you want to get this in first

File: torch/testing/_internal/common_methods_invocations.py (hunk 5 above)
Comment (soulitzer) [reply torch/testing/_internal/common_methods_invocations.py:45]: Good catch, it needs to be imported somewhere outside of testing. I keep forgetting to add this.

File: tools/autograd/derivatives.yaml (hunk 3 above)
Comment (soulitzer) [reply tools/autograd/derivatives.yaml:1972]: It's probably the case that forward AD makes the distinction between single element 1-D tensors and scalars, and vmap doesn't (if we expected vmap to make that distinction, then that would be a bug).

File: torch/testing/_internal/common_methods_invocations.py (hunk 4 above)
Comment (soulitzer) [reply torch/testing/_internal/common_methods_invocations.py:3330]: Whoops

File: torch/_decomp/decompositions_for_jvp.py (hunk 7 above)
Comment (samdow) [reply torch/_decomp/decompositions_for_jvp.py:182]: > native_layer_norm_backward's decomposition only gets registered to decomposition_table_for_jvp. Is there a reason why we didn't put it into torch._decomps.decomposition_table? @Chillee

This one is from me--(tl;dr, it's the norms issue). So there is a version of this that's basically the same in the `torch._decomps.decomposition_table` but the only difference is that this recomputes mean and var.

When running in eager at least, all norms end up in this state where the mean and var passed into the function don't have autograd tracking back to input. They assume that the backwards and double backwards kernels will know about this dependency since the kernels are specific to the norms. However, when we're doing decompositions, since we aren't writing a custom forwards over reverse kernel for the norm, we need to propagate this information to the so we end up recomputing versions of the mean and variance that do depend on the input. I didn't put this in the decompositions table because it really felt hacky. However, not sure if this is something that might come up with the normal decompositions (to be clear, the non-hacky version computes native_layer_norm_backward correctly but just doesn't do the forwards AD over that function correctly)

Comment (soulitzer) [reply torch/_decomp/decompositions_for_jvp.py:182]: (1) It does some recomputation for the purposes of higher-order gradient computation that we otherwise wouldn't need.
(2) Agreed. `clone` is probably a better choice here.

File: torch/_decomp/decompositions_for_jvp.py (hunk 6 above)
Comment (soulitzer) [reply torch/_decomp/decompositions_for_jvp.py:183]: The mechanism we are using to register decompositions doesn't seem to be exclusively used for jvp. So open question here is whether `torch/csrc/jit/runtime/decomposition_registry.cpp` is being used for other things.

Comment (zou3519) [reply torch/_decomp/decompositions_for_jvp.py:183]: Okay, that would make sense. I would put what you wrote in that code comment there so that future people reading the code who weren't around for the " special JVP_DECOMP DynamicLayerFront kernel" have better context

File: torch/_decomp/decompositions_for_jvp.py (hunk 7 above)
Comment (zou3519) [reply torch/_decomp/decompositions_for_jvp.py:182]: Ahh it's that. Thanks for the explanations

File: torch/_decomp/decompositions_for_jvp.py (hunk 6 above)
Comment (soulitzer) [reply torch/_decomp/decompositions_for_jvp.py:190]: Good idea, updated

File: torch/__init__.py [hunk 8]
```diff
@@ -892,6 +892,10 @@ def compiled_with_cxx11_abi():
 from torch._ops import ops
 from torch._classes import classes
 
+# decompositions_for_jvp on torch.ops
+# Registers decompositions for jvp to the jit registry
+from torch._decomp import decompositions_for_jvp
+
```
Comment (zou3519) [inline torch/__init__.py:898]: Does this make it public API? E.g. `torch.decompositions_for_jvp`. If so then we've got a problem

**Comment by zou3519** [comment]
> (2) I'd rather have this in a different PR since the changes required would probably be mostly unrelated to the ones here

SGTM

**Review by zou3519 — APPROVED** [review]
LGTM, but two comments:
- We should check that torch.decompositions_for_jvp isn't a user facing API (that would be bad)
- We should fix whatever is going on in vmapjvpall and log_sigmoid in a follow-up

File: torch/__init__.py (hunk 8 above)
Comment (soulitzer) [reply torch/__init__.py]: yes, nice catch - fixed

File: torch/__init__.py [hunk 9]
```diff
@@ -892,6 +892,12 @@ def compiled_with_cxx11_abi():
 from torch._ops import ops
 from torch._classes import classes
 
+# decompositions_for_jvp depends on torch.ops
+# Registers decompositions for jvp to the jit registry
+if os.environ.get("PYTORCH_JIT", "1") == "1":
+    from torch._decomp import decompositions_for_jvp
+    del decompositions_for_jvp
```
Comment (soulitzer) [inline torch/__init__.py:899]: Hey @eellison, any idea when `PYTORCH_JIT=0` is used? Should we be worried about users running into this and randomly having less coverage?

Comment (soulitzer) [reply torch/__init__.py:899]: This should probably be OK now:
- This shouldn't be a regression
- I've updated the error message now with `Note that forward AD support for some operators require JIT to be enabled. If the environment var PYTORCH_JIT=0 is set, some operators may no longer be used with forward AD`.

Comment (eellison) [reply torch/__init__.py:912]: I think it's extremely unlikely that will be used - tbh I forgot we have it and I don't know that we have a CI Job for it.

Comment (soulitzer) [reply torch/__init__.py:912]: Ah ok, thanks - there was a test that tries to import torch with PYTORCH_JIT=0 which is why I added this. I don't know if the message is super necessary anymore, but will keep for now.

**Comment by osalpekar** [comment]
This causes internal asan link-time failures: 
```
ld.lld: error: undefined symbol: torch::jit::has_jit_decomposition(c10::FunctionSchema const&)
>>> referenced by VariableTypeUtilsDependOnOps.h:22 (buck-out/gen/aab7ed39/xplat/caffe2/torch_headersAndroid#header-mode-symlink-tree-with-header-map,headers/torch/csrc/autograd/VariableTypeUtilsDependOnOps.h:22)
```
